File: frontend/src/api/datasets.ts
import { apiClient } from './client'
import type { DatasetPreview, DatasetSummary } from './types'

export async function listDatasets(): Promise<DatasetSummary[]> {
  const { data } = await apiClient.get<{ datasets: DatasetSummary[] }>('/datasets')
  return data.datasets
}

export interface UploadDatasetOptions {
  datasetName?: string
  plant?: string
  unit?: string
  onProgress?: (percent: number) => void
}

export async function uploadDataset(file: File, options: UploadDatasetOptions = {}): Promise<DatasetSummary> {
  const form = new FormData()
  form.append('file', file)
  if (options.datasetName) form.append('dataset_name', options.datasetName)
  if (options.plant) form.append('plant', options.plant)
  if (options.unit) form.append('unit', options.unit)
  // No explicit Content-Type for a FormData body — axios/the browser must
  // generate it (with the required multipart boundary) itself; overriding it
  // to a bare 'multipart/form-data' with no boundary produces a request the
  // server's multipart parser can't split into fields.
  const { data } = await apiClient.post<DatasetSummary>('/datasets/upload', form, {
    onUploadProgress: (e) => {
      if (options.onProgress && e.total) {
        options.onProgress(Math.round((e.loaded / e.total) * 100))
      }
    },
  })
  return data
}

export async function getDatasetPreview(name: string): Promise<DatasetPreview> {
  const { data } = await apiClient.get<DatasetPreview>(
    `/datasets/${encodeURIComponent(name)}/preview`,
  )
  return data
}

export async function deleteDataset(name: string): Promise<void> {
  await apiClient.delete(`/datasets/${encodeURIComponent(name)}`)
}

File: backend/app/api/routes/datasets.py
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Form, Query, UploadFile

from backend.app.schemas.datasets import DatasetListResponse, DatasetPreview, DatasetSummary
from backend.app.services import dataset_service

router = APIRouter(prefix="/datasets", tags=["datasets"])


@router.post("/upload", response_model=DatasetSummary, status_code=201)
async def upload_dataset(
    file: UploadFile,
    dataset_name: Optional[str] = Form(None),
    plant: Optional[str] = Form(None),
    unit: Optional[str] = Form(None),
    case_id: str = Query("default"),
) -> DatasetSummary:
    return await dataset_service.upload_dataset(
        file, dataset_name=dataset_name, plant=plant, unit=unit, case_id=case_id
    )


@router.get("", response_model=DatasetListResponse)
def list_datasets(case_id: str = Query("default")) -> DatasetListResponse:
    return DatasetListResponse(datasets=dataset_service.list_datasets(case_id))


@router.get("/{name}/preview", response_model=DatasetPreview)
def get_dataset_preview(name: str, case_id: str = Query("default")) -> DatasetPreview:
    return dataset_service.get_dataset_preview(name, case_id)


@router.delete("/{name}", status_code=204)
def delete_dataset(name: str, case_id: str = Query("default")) -> None:
    dataset_service.delete_dataset(name, case_id)
